source/common/router/header_parser.cc

```diff
@@ -214,45 +211,59 @@ HeaderFormatterPtr parseInternal(const envoy::config::core::v3::HeaderValue& hea
     return std::move(formatters[0]);
   }
 
-  return std::make_unique<CompoundHeaderFormatter>(std::move(formatters), append);
+  return std::make_unique<CompoundHeaderFormatter>(std::move(formatters));
 }
 
 } // namespace
 
-HeaderParserPtr HeaderParser::configure(
-    const Protobuf::RepeatedPtrField<envoy::config::core::v3::HeaderValueOption>& headers_to_add) {
+HeaderParserPtr
+HeaderParser::configure(const Protobuf::RepeatedPtrField<HeaderAppendOption>& headers_to_add) {
   HeaderParserPtr header_parser(new HeaderParser());
 
   for (const auto& header_value_option : headers_to_add) {
-    const bool append = PROTOBUF_GET_WRAPPED_OR_DEFAULT(header_value_option, append, true);
-    HeaderFormatterPtr header_formatter = parseInternal(header_value_option.header(), append);
+    HeaderAppendAction append_action;
+
+    if (header_value_option.has_append()) {
+      // 'append' is set and ensure the 'append_action' value is equal to the default value.
+      if (header_value_option.append_action() != HeaderAppendOption::APPEND_IF_EXISTS_OR_ADD) {
```

Comment:
This means if the user explicitly set `append` as true and `append_action` as APPEND_IF_EXISTS_OR_ADD is valid config.

But if we consider to deprecated `append` field, then this case should be invalid also. We encourage the user use the new field,  if both fields are specified, it should consider as invalid even if their value not conflict.

Reply:
I think there is no way to distinguish a explicit APPEND_IF_EXISTS_OR_ADD or default APPEND_IF_EXISTS_OR_ADD. 🤔

Reply:
emm...I thought there should be an interferce `has_append_action`, seems not.